lib/chat_web/controllers/zip_controller.ex

```diff
@@ -0,0 +1,138 @@
+defmodule ChatWeb.ZipController do
+  @moduledoc """
+  Enables downloading multiple messages in an archive.
+
+  The archive includes:
+  - index.html file showing all the selected messages
+  - files directory containing all the files from the selected messages
+  """
+
+  use ChatWeb, :controller
+
+  alias Chat.Rooms.Message
+  alias ChatWeb.MainLive.Layout.ExportedMessage
+  alias Chat.Broker
+  alias Chat.ChunkedFiles
+  alias Chat.Dialogs
+  alias Chat.Dialogs.PrivateMessage
+  alias Chat.Files
+  alias Chat.User
+  alias Chat.Utils.StorageId
+
+  def get(conn, params) do
+    with %{"broker_key" => broker_key} <- params,
+         {messages_ids, user_data, user_id, timezone} <- Broker.get(broker_key),
+         {identity, _rooms} = User.device_decode(user_data),
+         peer <- User.by_id(user_id),
+         dialog <- Dialogs.find_or_open(identity, peer),
+         {messages, users} when not is_nil(messages) and not is_nil(users) <-
+           fetch_messages_and_users_ids(dialog, identity, messages_ids) do
+      conn =
+        conn
+        |> put_resp_content_type("application/zip")
+        |> put_resp_header("content-disposition", "attachment; filename=messages.zip")
+        |> send_chunked(200)
+
+      style_stream =
+        :chat
+        |> :code.priv_dir()
+        |> Path.join("static/assets/app.css")
+        |> File.stream!()
+
+      messages_stream =
+        messages
+        |> Stream.map(fn message ->
+          author =
+            case message do
+              %PrivateMessage{} ->
+                peer
+
+              %Message{} = message ->
+                Map.get(users, message.author_hash)
+            end
+
+          %{author: author, message: message, timezone: timezone}
+          |> ExportedMessage.message_block()
+          |> Phoenix.HTML.Safe.to_iodata()
+        end)
+
+      exported_messages =
+        Stream.concat([
+          ["<html><head>"],
+          ["<style>"],
+          style_stream,
+          ["</style></head><body>"],
+          messages_stream,
+          ["</body></html>"]
+        ])
+
+      index_entry = Zstream.entry("index.html", exported_messages)
+
+      file_entries =
+        messages
+        |> Enum.reduce([], fn message, file_entries ->
+          with %{type: type, content: json} when type in [:file, :image, :video] <- message,
+               {id, content} <- StorageId.from_json(json),
+               [chunk_key, chunk_secret_raw, _, _type, filename, _size] <- Files.get(id, content),
+               chunk_secret <- Base.decode64!(chunk_secret_raw),
+               file_stream <- ChunkedFiles.stream_chunks(chunk_key, chunk_secret) do
```

Comment:
`file_stream` is a binary with whole file, not a stream. This needs to be stream to make memory footprint not to reflect file size. Files could be huge. Target system may have 2G of physical RAM only

Reply:
File chunk go back to db. So there would be `CubDB.Select` to get a scream of chunks.
Check the https://github.com/Buckitup-chat/chat/pull/63 or wait till it gets into master. Btw, any critics is appreciative